bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyScrollingControl.java

```diff
@@ -0,0 +1,482 @@
+			extension.removeTextPresentationListener(textPresentationListener);
+		}
+		if (sourceViewer.getTextWidget() != null) {
+			sourceViewer.getTextWidget().removeControlListener(controlListener);
+			sourceViewer.getTextWidget().removeKeyListener(caretListener);
+			sourceViewer.getTextWidget().removeCaretListener(caretListener);
+		}
+		this.stickyLinesCanvas.dispose();
+	}
+
+	private void createControls() {
+		Composite sourceViewerComposite= null;
+		if (sourceViewer instanceof ITextViewerExtension extension) {
+			sourceViewerComposite= (Composite) extension.getControl();
+		} else {
+			sourceViewerComposite= sourceViewer.getTextWidget().getParent();
+		}
+
+		stickyLinesCanvas= new Canvas(sourceViewerComposite, SWT.NONE);
+		addMouseListeners(stickyLinesCanvas);
+		GridLayoutFactory.fillDefaults().numColumns(2).spacing(0, 0).applyTo(stickyLinesCanvas);
+
+		stickyLineNumber= new StyledText(stickyLinesCanvas, SWT.LEFT | SWT.WRAP);
+		GridDataFactory.fillDefaults().grab(false, true).exclude(verticalRuler == null).applyTo(stickyLineNumber);
+		stickyLineNumber.setVisible(verticalRuler != null);
+		stickyLineNumber.setEnabled(false);
+		stickyLineNumber.setBackground(settings.stickyLineBackgroundColor());
+
+		stickyLineText= new StyledText(stickyLinesCanvas, SWT.NONE);
+		GridDataFactory.fillDefaults().grab(true, true).applyTo(stickyLineText);
+		stickyLineText.setEnabled(false);
+		stickyLineText.setBackground(settings.stickyLineBackgroundColor());
+
+		bottomSeparator= new Label(stickyLinesCanvas, SWT.SEPARATOR | SWT.SHADOW_OUT | SWT.HORIZONTAL);
+		GridDataFactory.fillDefaults().grab(true, false).span(2, 1).applyTo(bottomSeparator);
+		bottomSeparator.setEnabled(false);
+
+		bottomSeparator= new Label(stickyLinesCanvas, SWT.SEPARATOR | SWT.SHADOW_OUT | SWT.HORIZONTAL);
+		GridDataFactory.fillDefaults().grab(true, false).indent(0, 2).span(2, 1).applyTo(bottomSeparator);
+		bottomSeparator.setEnabled(false);
+
+		stickyLinesCanvas.moveAbove(null);
+	}
+
+	private void updateStickyScrollingControls() {
+		StringJoiner stickyLineTextJoiner= new StringJoiner(System.lineSeparator());
+		StringJoiner stickyLineNumberJoiner= new StringJoiner(System.lineSeparator());
+		for (int i= 0; i < getNumberStickyLines(); i++) {
+			StickyLine stickyLine= stickyLines.get(i);
+			stickyLineTextJoiner.add(stickyLine.text());
+			stickyLineNumberJoiner.add(fillLineNumberWithLeadingSpaces(stickyLine.lineNumber() + 1));
+		}
+
+		String newStickyLineText= stickyLineTextJoiner.toString();
+		String newStickyLineNumber= stickyLineNumberJoiner.toString();
+
+		if (!newStickyLineText.equals(stickyLineText.getText())
+				|| !newStickyLineNumber.equals(stickyLineNumber.getText())) {
+			stickyLineText.setText(stickyLineTextJoiner.toString());
+			stickyLineNumber.setText(stickyLineNumberJoiner.toString());
+			styleStickyLines();
+			layoutStickyLines();
+		}
+	}
+
+	private String fillLineNumberWithLeadingSpaces(int lineNumber) {
```

Comment:
Isn't there a way to use an [alignment option of SWT](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fswt%2Fcustom%2FStyledText.html)?

Reply:
I tried to implement right alignment but ended up mirroring the logic used in the line number ruler, as it also lacks specified alignment. (see [LineNumberColumn.java](https://github.com/eclipse-platform/eclipse.platform.ui/blob/385b30bb74bce1bc048584281a7b4fdedc028456/bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/source/LineNumberRulerColumn.java#L546C2-L546C11)).